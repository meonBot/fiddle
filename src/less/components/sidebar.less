#new-file-input {
  box-shadow: 0 0 0 1600px rgba(0, 0, 0, 0.1);
  border-radius: 0;
  border-left: 4px solid white;
  font-size: 14px;
}

.add-file-input .bp3-tree-node-label {
  overflow: visible;
}

.bp3-tree-node-caret-none {
  min-width: 8px;
}

.bp3-tree-node-content-1 {
  padding-left: 0;
}
