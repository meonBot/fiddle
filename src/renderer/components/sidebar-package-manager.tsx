import * as React from 'react';
import { Button, InputGroup, Tree, TreeNodeInfo } from '@blueprintjs/core';
import { AppState } from '../state';
import { observer } from 'mobx-react';

interface IState {
  search: string;
}

interface IProps {
  appState: AppState;
}

@observer
export class SidebarPackageManager extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      search: '',
    };
  }

  public addPackageToSearch = (event: React.KeyboardEvent) => {
    const { search } = this.state;
    const { appState } = this.props;
    if (event.key === 'Enter' && !!search) {
      // TODO: allow users to specify the package version
      appState.modules.set(search, '*');
      this.setState({ search: '' });
    }
  };

  public render() {
    return (
      <Tree
        contents={[
          {
            childNodes: [
              {
                id: '-1',
                label: (
                  <InputGroup
                    leftIcon="search"
                    placeholder="Add package"
                    fill
                    onChange={(event) =>
                      this.setState({ search: event.target.value })
                    }
                    onKeyPress={this.addPackageToSearch}
                    value={this.state.search}
                  />
                ),
              },
              ...this.renderModules(),
            ],
            id: 'modules',
            hasCaret: false,
            icon: 'code-block',
            label: 'Modules',
            isExpanded: true,
          },
        ]}
      />
    );
  }

  public renderModules = (): TreeNodeInfo[] => {
    const values: TreeNodeInfo[] = [];
    const { appState } = this.props;
    for (const pkg of appState.modules.keys()) {
      values.push({
        id: pkg,
        label: pkg,
        secondaryLabel: (
          <Button
            minimal
            icon="remove"
            onClick={() => appState.modules.delete(pkg)}
          />
        ),
      });
    }

    return values;
  };
}
