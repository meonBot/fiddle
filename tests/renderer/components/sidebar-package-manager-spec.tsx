import { Button } from '@blueprintjs/core';
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { SidebarPackageManager } from '../../../src/renderer/components/sidebar-package-manager';

describe('SidebarPackageManager component', () => {
  let store: any;
  beforeEach(() => {
    store = {
      modules: new Map<string, string>([['cow', '*']]),
    };
  });

  it('renders', () => {
    const wrapper = shallow(<SidebarPackageManager appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('can add a package', () => {
    const wrapper = shallow(<SidebarPackageManager appState={store} />);
    const instance = wrapper.instance();
    instance.state = {
      search: 'say',
    };

    (instance as any).addPackageToSearch({
      key: 'Enter',
    });

    expect(
      Array.from((store.modules as Map<string, string>).entries()),
    ).toMatchSnapshot([
      ['cow', '*'],
      ['say', '*'],
    ]);
  });

  it('can remove a package', () => {
    const wrapper = mount(<SidebarPackageManager appState={store} />);
    const instance = wrapper.instance();
    instance.state = {
      search: 'say',
    };

    const btn = wrapper.find(Button);
    btn.simulate('click');
    expect((store.modules as Map<string, string>).size).toBe(0);
  });
});
