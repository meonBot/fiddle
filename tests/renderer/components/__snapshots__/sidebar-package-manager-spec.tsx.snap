// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarPackageManager component can add a package 1`] = `
Array [
  Array [
    "cow",
    "*",
  ],
  Array [
    "say",
    "*",
  ],
]
`;

exports[`SidebarPackageManager component renders 1`] = `
<Blueprint3.Tree
  contents={
    Array [
      Object {
        "childNodes": Array [
          Object {
            "id": "-1",
            "label": <Blueprint3.InputGroup
              fill={true}
              leftIcon="search"
              onChange={[Function]}
              onKeyPress={[Function]}
              placeholder="Add package"
              value=""
            />,
          },
          Object {
            "id": "cow",
            "label": "cow",
            "secondaryLabel": <Blueprint3.Button
              icon="remove"
              minimal={true}
              onClick={[Function]}
            />,
          },
        ],
        "hasCaret": false,
        "icon": "code-block",
        "id": "modules",
        "isExpanded": true,
        "label": "Modules",
      },
    ]
  }
/>
`;
